text/0000-export.md

```diff
@@ -0,0 +1,398 @@
+### Privacy
+
+It is an error to export an item that is not public, or is part of a non-public module.
+The set of exported items of a crate will always be a subset of the crate's public interface.
+
+It's fine to `#[export]` a public alias of a public type from a private module:
+
+```rust
+mod a {
+    pub extern "C" fn f() { … }
+}
+
+#[export]
+pub mod b {
+    pub use super::a::f;
+}
+```
+
+(This will export the function f as `b::f`.)
+
+### Importing Exported Items
+
+Normally, when using a crate as a dependency, any `#[export]` attributes of that crate have no effect
+and the dependency is statically linked into the resulting binary.
+
+When explicitly specifying `dynamic = true` for the dependency with `Cargo.toml`,
+or when using a `extern dyn crate …;` statement in the source code,
+only the items marked as `#[export]` will be available and the dependency will be linked dynamically.
+
+### Building Dynamic Dependencies
+
+When using `dynamic = true` for a dependency, there is no need to build that full crate:
+only the signatures of its exported items are necessary.
+Cargo will pass a flag to the Rust compiler which will stop it from generating
+code for non-exported items and function bodies.
+
+A clear separation between "public dependencies" (which used in the interface)
+and "private dependencies" (which are only used in the implementation) is required
+to avoid building unnecessary indirect dependencies.
+A system for that has been proposed in [RFC 1977](https://rust-lang.github.io/rfcs/1977-public-private-dependencies.html).
+
+### Name Mangling and Safety
+
+Because a dynamic dependency and the crate that uses it are compiled separately
+and only combined at runtime,
+it is impossible for the compiler to perform any (safety, borrow, signature, …) checks.
+However, making a (perhaps accidental) change to a function signature or type
+should not lead to undefined behavior at runtime.
+
+There are two ways to solve this problem:
+
+1. Make it the responsibility of the user.
+2. Make it the responsibility of the loader/linker.
+
+Option (1) simply means making everything `unsafe`, which isn't very helpful to the user.
+Option (2) means the loader (the part that loads the dynamic library at runtime) needs to perform the checks.
+
+Unless we ship our own loader as part of Rust binaries,
+we can only make use of the one functionality available in the loaders of all operating systems:
+looking up symbols by their name.
+
+So, in order to be able to provide safety, the symbol name has to be unique for the full signature,
+including all relevant type descriptions.
+
+To avoid extremely long symbol names that contain a full (encoded) version of the function signature
+and all relevant type descriptions, we use a 128-bit hash based on all this information.
+
+For example, an exported item in `foo::bar` in the crate `mycrate` would be exported with a symbol name such as:
+
+```
+_RNvNtC_7mycrate3foo3bar_f8771d213159376fafbff1d3b93bb212
+```
```

Comment:
This doesn't seem to be able to handle multiple semver incompatible versions of the same crate. We currently include the `StableCrateId` in the symbol name for this which includes all `-Cmetadata` arguments. Cargo uses a separate `-Cmetadata` value for every crate version.

Reply:
I'm not sure if we can properly support that. If e.g. `libmycrate.so` is separately compiled, we cannot (and should not) make the symbols dependent on the crate id it will have in a dependency tree of whatever binary will use this library later.

Perhaps we want to include just the major version in the symbol names, but not everyone uses versions (a cargo concept) or follows semver, nor does the version of the public *API* need to match the public of the stable *ABI*.

Maybe we need an crate-global attribute to optionally specify the ABI version (or even just a random hash) for the crate, to allow for symbols with identical paths and signatures of multiple identically named crates to exist at once, although I imagine most use cases would not use e.g. two different versions of `librustls.so` at the same time. (If that's a supported use case, they should probably be called `librustls1.so` and `librustls2.so`, using `rustls1` and `rustls2` as the name used in the symbols rather than just `rustls`.)

Reply:
The `StableCrateId` does not depend on any dependencies. It only depends on the crate name, if it is a lib or bin and all `-Cmetadata` arguments (in sorted order). The SVH (crate hash) does depend on dependencies. In addition the current cargo impl hashes dependency versions into the `-Cmetadata` argument it passes, but that can easily be changed to only pass the major part of the semver version.

Reply:
That could work, but I'm not convinced the API version and the ABI version should be treated as one and the same. Renaming structs or non-exported pub functions/methods etc. etc. could result in an incompatible API while preserving a compatible ABI.

As I mentioned above, it seems a weird situation if a binary somehow ends up depending on two different `librustls.so`, considering they'd both have to have the same file name. (And if those are called `librustls1.so` and `librustls2.so`, then the `rustls1` and `rustls2` names are probably what should be used in the symbol names rather than just `rustls`. And that could just be a build option to cargo or a property in Cargo.toml.)

Reply:
These arguments actually bring into question whether there should be a supported ABI version in Cargo as well. Since Cargo *already* reconciles latest-compatible updates for APIs, why shouldn't it do the same with ABIs too?